06x-new-kafka-roller.md

```diff
@@ -0,0 +1,220 @@
+       3. `BROKER_AND_NOT_ACTIVE_CONTROLLER` - Node that is at least a broker but also might be a controller (combined) and is not the active controller
+       4. `BROKER_AND_ACTIVE_CONTROLLER` - Combined node that is the active controller (the quorum leader)
+
+       The returned batch will contain only one node if it is not `BROKER_AND_NOT_ACTIVE_CONTROLLER` group, so that controllers are restarted one at a time.
+  
+   - If `NON_ACTIVE_PURE_CONTROLLER` group is non empty, return the first node that can be restarted without impacting the quorum health (more on this later).
+   - If `ACTIVE_PURE_CONTROLLER` group is non empty, return the node if it can be restarted without impacting the quorum health. Otherwise return an empty set.
+   - If `BROKER_AND_NOT_ACTIVE_CONTROLLER` group is non empty, batch the broker nodes:
+       - remove the node from the list, if it is a combined node and cannot be restarted without impacting the quorum health so that it does get included in a batch
+       - build a map of nodes and their replicating partitions by sending describeTopics request to Admin API
+       - batch the nodes that do not have any partitions in common therefore can be restarted together
+       - remove nodes that have an impact on the availability from the batches (more on this later)
+       - return the largest batch
+   - If an empty batch is returned, that means none of the nodes met the safety conditions such as availability and qourum health impact. In this case, check their `numAttempts` and if any of them is equal to or greater than `maxAttempts`, throw `UnrestartableNodesException`. Otherwise increment their `numAttempts` and restart from step 3.
+
+8. Restart the nodes from the returned batch in parallel:
+   - If `numRestarts` of a node is larger than `maxRestarts`, throw `MaxRestartsExceededException`.
+   - Otherwise, restart each node and transition its state to `RESTARTED` and increment its `numRestarts`.
+   - After restarting all the nodes in the batch, wait for their states to become `SERVING` until the configured `postOperationalTimeoutMs` is reached.
+   -  If the timeout is reached, throw `TimeoutException` if a node's `numAttempts` is greater than or equal to `maxAttempts`. Otherwise increment their `numAttempts` and restart from step 3. 
+
+9. If there are no exceptions thrown at this point, the reconciliation completes successfully. If there were `UnrestartableNodesException`, `TimeoutException`, `MaxRestartsExceededException` or any other unexpected exceptions throws, the reconciliation fails.
+
+TODO: we need to figure out when to elect preferred leaders and not fail the reconciliation if did not become leaders within the timeout. This does not apply to pure controllers.
+
+#### Restarting not running combined nodes
+
+When restarting not running combined nodes, we will apply a special logic to address the issue described in https://github.com/strimzi/strimzi-kafka-operator/issues/9426.
```

Comment:
"restarting not running" sounds weird: You can't restart something that's not started. Perhaps we need a better form of works than "not running", wdyt?

Reply:
I removed this section because it looks repeated now that I added more information in the algorithm step.